node/chainSimulator/process/processor.go

```diff
@@ -106,17 +108,97 @@ func (creator *blocksCreator) CreateNewBlock() error {
 		return err
 	}
 
+	notNil := !check.IfNil(currentHeader)
+	enableEpochHandler := creator.nodeHandler.GetCoreComponents().EnableEpochsHandler()
+	var previousProof *dataBlock.HeaderProof
+	if notNil && enableEpochHandler.IsFlagEnabled(common.EquivalentMessagesFlag) {
+		sig, errS := creator.generateSignature(prevHash, blsKey.PubKey(), currentHeader)
+		if errS != nil {
+			return errS
+		}
+
+		previousProof = &dataBlock.HeaderProof{
+			PubKeysBitmap:       pubKeyBitmap,
+			AggregatedSignature: sig,
+			HeaderHash:          prevHash,
+			HeaderEpoch:         currentHeader.GetEpoch(),
+			HeaderNonce:         currentHeader.GetNonce(),
+			HeaderShardId:       currentHeader.GetShardID(),
+			HeaderRound:         currentHeader.GetNonce(),
+			IsStartOfEpoch:      currentHeader.IsStartOfEpochBlock(),
+		}
+
+		wasAdded := creator.nodeHandler.GetDataComponents().Datapool().Proofs().AddProof(previousProof)
+		if !wasAdded {
+			log.Warn("proofsPool.AddProof not added", "currHdrHash", prevHash)
+		}
+	}
+
 	header, block, err := bp.CreateBlock(newHeader, func() bool {
 		return true
 	})
 	if err != nil {
 		return err
 	}
 
-	err = creator.setHeaderSignatures(header, blsKey.PubKey())
+	if notNil && common.ShouldBlockHavePrevProof(header, enableEpochHandler, common.EquivalentMessagesFlag) {
+		selectionEpoch := epoch
+		if currentHeader.IsStartOfEpochBlock() {
+			selectionEpoch = selectionEpoch - 1
+		}
+
+		_, validators, err = creator.nodeHandler.GetProcessComponents().NodesCoordinator().ComputeConsensusGroup(currentHeader.GetPrevRandSeed(), currentHeader.GetRound(), shardID, selectionEpoch)
+		if err != nil {
+			return err
+		}
+
+		previousProof.PubKeysBitmap = GeneratePubKeyBitmap(len(validators))
+		previousProof.AggregatedSignature, err = creator.generateSignatureForProofs(prevHash, previousProof, validators)
+		if err != nil {
+			return err
+		}
+
+		header.SetPreviousProof(previousProof)
+	}
+
+	err = creator.setHeaderSignatures(header, blsKey.PubKey(), validators)
+	if err != nil {
+		return err
+	}
+
+	marshalizedHdr, err := creator.nodeHandler.GetCoreComponents().
+		InternalMarshalizer().Marshal(header)
 	if err != nil {
 		return err
 	}
+	headerHash := creator.nodeHandler.GetCoreComponents().Hasher().Compute(string(marshalizedHdr))
+
+	pubKeys := extractValidatorPubKeys(validators)
+	newHeaderSig, err := creator.generateAggregatedSignature(headerHash, header.GetEpoch(), header.GetPubKeysBitmap(), pubKeys)
+	if err != nil {
+		return err
+	}
+
+	var headerProof *dataBlock.HeaderProof
```

Comment:
this could be extracted into a func createAndAddCurrentProofToPool

Reply:
refactored